### PR TITLE
Use type_name/type_namespace for inline named types

### PR DESCRIPTION
**Pull request by tjwp**
The original premise for this change was to use the options `type_name` and `type_namespace` to name  types defined inline.

The actual implementation became more complex in providing good feedback when the wrong attribute name is used, or used in the wrong location.

There are two commits here: 
- the first is a rename of the internal `type_name` to `avro_type_name`. 
- the second has all the changes to support `type_name`/`type_namespace` as options and the error handling.

Examples:

```ruby
# Top-level examples

# GOOD
record :foo

# GOOD
record :foo, namespace: :bar

# GOOD
record :foo, aliases: :bar

# BAD
record name: :foo

# BAD
record :foo do
  namespace :bar
end

# BAD
record do
  name: :foo
end

# BAD
record type_name: :foo

# BAD
record do
  type_name :foo
end

# BAD
record :foo, type_namespace :bar

# BAD
record :foo do
  type_namespace :bar
end

# BAD
record :foo, type_aliases: :bar

# Inline examples
# all of these apply to :record, :fixed and :enum

# GOOD
record :rec do
  required :sub, :record, type_name: :foo
end

# GOOD
record :rec do
  required :sub, :record, type_namespace: :bar
end

# GOOD
record :rec do
  required :sub, :record, type_name: :foo, type_aliases: :baz
end

# GOOD
record :rec do
  required :sub, :record, do
    type_aliases :baz
  end
end

# BAD
record :rec do
  required :sub, :record, name: :foo
end

# BAD
record :rec do
  required :sub, :record do
    type_name :foo
  end
end

# BAD
record :rec do
  required :sub, :record, namespace: :bar
end

# BAD
record :rec do
  required :sub, :record do
    type_namespace: :bar
  end
end
```

Prime: @jturkel

### CODE REVIEW

File: lib/avro/builder/dsl.rb [hunk 1]
```diff
@@ -60,6 +63,7 @@ def to_h
 
       # Return the last schema object processed as an Avro JSON schema
       def to_json(validate: true, pretty: true)
+
```
Comment (tjwp) [inline lib/avro/builder/dsl.rb]: I'll remove this whitespace line.

File: lib/avro/builder/dsl_options.rb [hunk 2]
```diff
@@ -19,11 +19,17 @@ def dsl_option?(name)
       end
 
       module ClassMethods
-        # A DSL option is only settable as an option, not as method in a block.
-        def dsl_option(name, &block)
+
+        # Defines a private writer with new_name= to set an attribute stored in the
```
Comment (tjwp) [inline lib/avro/builder/dsl_options.rb]: This should say "dsl_name=".

**Comment by jturkel** [comment]
:shipit:

**Comment by tjwp** [comment]
@jturkel That was easier than I expected! I hope I haven't induced PR fatigue.

**Pull request closed without merging**